File: app/src/main/java/org/wikipedia/page/LinkMovementMethodExt.java
package org.wikipedia.page;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.Layout;
import android.text.Spannable;
import android.text.method.LinkMovementMethod;
import android.text.style.URLSpan;
import android.view.MotionEvent;
import android.widget.TextView;

import org.wikipedia.util.UriUtil;
import org.wikipedia.util.log.L;

import static org.wikipedia.util.UriUtil.decodeURL;

/**
 * Intercept web links and add special behavior for external links.
 */
public class LinkMovementMethodExt extends LinkMovementMethod {
    @Nullable private UrlHandler handler;
    @Nullable private UrlHandlerWithText handlerWithText;

    public LinkMovementMethodExt(@Nullable UrlHandler handler) {
        this.handler = handler;
    }

    public LinkMovementMethodExt(@Nullable UrlHandlerWithText handler) {
        this.handlerWithText = handler;
    }

    @Override
    public boolean onTouchEvent(@NonNull final TextView widget, @NonNull final Spannable buffer, @NonNull final MotionEvent event) {
        final int action = event.getAction();
        if (action == MotionEvent.ACTION_UP) {
            final int x = (int) event.getX() - widget.getTotalPaddingLeft() + widget.getScrollX();
            final int y = (int) event.getY() - widget.getTotalPaddingTop() + widget.getScrollY();
            final Layout layout = widget.getLayout();
            final int line = layout.getLineForVertical(y);
            final int off = layout.getOffsetForHorizontal(line, x);
            final URLSpan[] links = buffer.getSpans(off, off, URLSpan.class);
            if (links.length != 0) {
                String linkText;
                try {
                    linkText = buffer.subSequence(buffer.getSpanStart(links[0]), buffer.getSpanEnd(links[0])).toString();
                } catch (Exception e) {
                    e.printStackTrace();
                    linkText = "";
                }
                L.d(linkText);
                String url = decodeURL(links[0].getURL());
                if (handler != null) {
                    handler.onUrlClick(url);
                } else if (handlerWithText != null) {
                    handlerWithText.onUrlClick(url, UriUtil.getTitleFromUrl(url), linkText);
                }
                return true;
            }
        }
        return super.onTouchEvent(widget, buffer, event);
    }

    public interface UrlHandler {
        void onUrlClick(@NonNull String url);
    }

    public interface UrlHandlerWithText {
        void onUrlClick(@NonNull String url, @Nullable String titleString, @NonNull String linkText);
    }
}

File: app/src/main/java/org/wikipedia/page/LinkMovementMethodExt.kt
package org.wikipedia.page

import android.net.Uri
import android.text.Spannable
import android.text.method.LinkMovementMethod
import android.text.style.URLSpan
import android.view.MotionEvent
import android.widget.TextView
import androidx.core.text.getSpans
import org.wikipedia.WikipediaApp
import org.wikipedia.util.UriUtil
import org.wikipedia.util.log.L

class LinkMovementMethodExt : LinkMovementMethod {

    fun interface UrlHandler {
        fun onUrlClick(url: String)
    }

    fun interface UrlHandlerWithText {
        fun onUrlClick(url: String, titleString: String?, linkText: String)
    }

    fun interface UrlHandlerWithTextAndCoords {
        fun onUrlClick(url: String, titleString: String?, linkText: String, x: Int, y: Int)
    }

    private var handler: UrlHandler? = null
    private var handlerWithText: UrlHandlerWithText? = null
    private var handlerWithTextAndCoords: UrlHandlerWithTextAndCoords? = null

    constructor(handler: UrlHandler?) {
        this.handler = handler
    }

    constructor(handler: UrlHandlerWithText?) {
        handlerWithText = handler
    }

    constructor(handler: UrlHandlerWithTextAndCoords?) {
        handlerWithTextAndCoords = handler
    }

    override fun onTouchEvent(widget: TextView, buffer: Spannable, event: MotionEvent): Boolean {
        val action = event.action
        if (action == MotionEvent.ACTION_UP) {
            val x = event.x.toInt() - widget.totalPaddingLeft + widget.scrollX
            val y = event.y.toInt() - widget.totalPaddingTop + widget.scrollY
            val layout = widget.layout
            val line = layout.getLineForVertical(y)
            val off = layout.getOffsetForHorizontal(line, x.toFloat())
            val links = buffer.getSpans<URLSpan>(off, off)
            if (links.isNotEmpty()) {
                val linkText = try {
                    buffer.subSequence(buffer.getSpanStart(links[0]), buffer.getSpanEnd(links[0])).toString()
                } catch (e: Exception) {
                    e.printStackTrace()
                    ""
                }
                L.d(linkText)
                val url = UriUtil.decodeURL(links[0].url)

                handler?.run {
                    onUrlClick(url)
                }

                handlerWithText?.run {
                    onUrlClick(url, UriUtil.getTitleFromUrl(url), linkText)
                }

                handlerWithTextAndCoords?.run {
                    onUrlClick(url, UriUtil.getTitleFromUrl(url), linkText, event.rawX.toInt(), event.rawY.toInt())
                }

                return true
            }
        }
        return super.onTouchEvent(widget, buffer, event)
    }

    internal class ErrorLinkHandler internal constructor() : LinkHandler(WikipediaApp.getInstance()) {
        override var wikiSite = WikipediaApp.getInstance().wikiSite
        override fun onMediaLinkClicked(title: PageTitle) {}
        override fun onDiffLinkClicked(title: PageTitle, revisionId: Long) {}
        override fun onPageLinkClicked(anchor: String, linkText: String) {}
        override fun onInternalLinkClicked(title: PageTitle) {
            // Explicitly send everything to an external browser, since the error might be shown in
            // a child activity of PageActivity, and we don't want to lose our place.
            UriUtil.visitInExternalBrowser(WikipediaApp.getInstance(),
                    Uri.parse(UriUtil.resolveProtocolRelativeUrl(title.wikiSite, title.mobileUri)))
        }
    }

    companion object {
        fun getExternalLinkMovementMethod(): LinkMovementMethodExt {
            return LinkMovementMethodExt(ErrorLinkHandler())
        }
    }
}
